include/deal.II/base/symmetric_tensor.h

```diff
@@ -2320,6 +2325,1003 @@ Number second_invariant (const SymmetricTensor<2,3,Number> &t)
+      Number norm;          // Squared norm or inverse norm of current eigenvector
+      Number t, u;          // Intermediate storage
+
+      // Calculate eigenvalues
+      const std_cxx11::array<Number,dim> w = eigenvalues(A);
+
+      t = std::abs(w[0]);
+      for (unsigned int i=1; i<dim; ++i)
+        {
+          u = std::abs(w[i]);
+          if (u > t)
+            t = u;
+        }
+
+      if (t < 1.0)
+        u = t;
+      else
+        u = t*t;
+
+      // Estimated maximum roundoff error
+      const double error = 256.0 * std::numeric_limits<double>::epsilon() * u*u;
+
+      // Initialize the transformation matrix as the
+      // identity tensor
+      dealii::Tensor<2,dim,Number> Q;
+      Q[0][1] = A[0][1]*A[1][2] - A[0][2]*A[1][1];
+      Q[1][1] = A[0][2]*A[0][1] - A[1][2]*A[0][0];
+      Q[2][1] = A[0][1]*A[0][1];
+
+      // Calculate first eigenvector by the formula
+      //   v[0] = (A - w[0]).e1 x (A - w[0]).e2
+      Q[0][0] = Q[0][1] + A[0][2]*w[0];
+      Q[1][0] = Q[1][1] + A[1][2]*w[0];
+      Q[2][0] = (A[0][0] - w[0]) * (A[1][1] - w[0]) - Q[2][1];
+      norm    = Q[0][0]*Q[0][0] + Q[1][0]*Q[1][0] + Q[2][0]*Q[2][0];
+
+      // If vectors are nearly linearly dependent, or if there might have
+      // been large cancellations in the calculation of A[i][i] - w[0], fall
+      // back to QL algorithm
+      // Note that this simultaneously ensures that multiple eigenvalues do
+      // not cause problems: If w[0] = w[1], then A - w[0] * I has rank 1,
+      // i.e. all columns of A - w[0] * I are linearly dependent.
+      if (norm <= error)
+        {
+          return ql_implicit_shifts(A);
+        }
+      else                      // This is the standard branch
+        {
+          norm = std::sqrt(1.0 / norm);
+          for (unsigned j=0; j < dim; j++)
+            Q[j][0] = Q[j][0] * norm;
+        }
+
+      // Calculate second eigenvector by the formula
+      //   v[1] = (A - w[1]).e1 x (A - w[1]).e2
+      Q[0][1]  = Q[0][1] + A[0][2]*w[1];
+      Q[1][1]  = Q[1][1] + A[1][2]*w[1];
+      Q[2][1]  = (A[0][0] - w[1]) * (A[1][1] - w[1]) - Q[2][1];
+      norm     = Q[0][1]*Q[0][1] + Q[1][1]*Q[1][1] + Q[2][1]*Q[2][1];
+      if (norm <= error)
+        {
+          return ql_implicit_shifts(A);
+        }
+      else
+        {
+          norm = std::sqrt(1.0 / norm);
+          for (unsigned int j=0; j < dim; j++)
+            Q[j][1] = Q[j][1] * norm;
+        }
+
+      // Calculate third eigenvector according to
+      //   v[2] = v[0] x v[1]
+      Q[0][2] = Q[1][0]*Q[2][1] - Q[2][0]*Q[1][1];
+      Q[1][2] = Q[2][0]*Q[0][1] - Q[0][0]*Q[2][1];
+      Q[2][2] = Q[0][0]*Q[1][1] - Q[1][0]*Q[0][1];
+
+      // Structure the data to be outputted
+      std_cxx11::array<std::pair<Number, Tensor<1,dim,Number> >,dim> eig_vals_vecs;
+      for (unsigned int e=0; e<dim; ++e)
+        {
+          eig_vals_vecs[e].first = w[e];
+
+          // The column "e" of Q contains the non-normalised
```

Comment:
Drats, I thought I'd tracked down all of my Britishisms :-P Thanks for pointing all of these out for me!